pennylane/templates/layers/random.py

```diff
@@ -99,10 +94,116 @@ def RandomLayers(weights, wires, ratio_imprim=0.3, imprimitive=CNOT, rotations=N
+
+        >>> np.allclose(circuit1(weights), circuit2(weights))
+        >>> True
+
+        You can verify this by drawing the circuits.
+
+            >>> print(circuit1.draw())
+            >>>  0: ──RX(0.1)──RX(-2.1)──╭X──╭X───────────┤ ⟨Z⟩
+            ...  1: ─────────────────────╰C──╰C──RZ(1.4)──┤
+
+            >>> print(circuit2.draw())
+            >>>  0: ──RX(0.1)──RX(-2.1)──╭X──╭X───────────┤ ⟨Z⟩
+            ...  1: ─────────────────────╰C──╰C──RZ(1.4)──┤
+
+        **Changing the seed**
+
+        To change the randomly generated circuit architecture, you have to change the seed passed to the template.
+        For example, these two calls of ``RandomLayers`` *do not* create the same circuit:
+
+        .. code-block:: python
+
+            @qml.qnode(dev)
+            def circuit_9(weights):
+                RandomLayers(weights=weights, wires=range(2), seed=9)
+                return qml.expval(qml.PauliZ(0))
+
+            @qml.qnode(dev)
+            def circuit_12(weights):
+                RandomLayers(weights=weights, wires=range(2), seed=12)
+                return qml.expval(qml.PauliZ(0))
+
+        >>> np.allclose(circuit_9(weights), circuit_12(weights))
+        >>> False
+
+        >>> print(circuit_9.draw())
+        >>>  0: ──╭X──RY(-2.1)──RX(1.4)──┤ ⟨Z⟩
+        ...  1: ──╰C──RX(0.1)────────────┤
+
+        >>> print(circuit_12.draw())
+        >>>  0: ──╭X──RX(-2.1)──╭C──╭X──RZ(1.4)──┤ ⟨Z⟩
+        ...  1: ──╰C──RZ(0.1)───╰X──╰C───────────┤
+
+
+        **Automatically creating random circuits**
+
+        To automate the process of creating different circuits with ``RandomLayers``,
+        you can pass a random number generator as seed. However, in this case care needs
+        to be taken. In the default setting, a quantum node is **mutable**, which means that the quantum function is
+        re-evaluated every time it is called. This means that the circuit is re-constructed from scratch
+        each time you call the qnode:
+
+        .. code-block:: python
+
+            @qml.qnode(dev)
+            def circuit_rnd(weights):
+                RandomLayers(weights=weights, wires=range(2), seed=np.random.randint(100))
+                return qml.expval(qml.PauliZ(0))
+
+            first_call = circuit_rnd(weights)
+            second_call = circuit_rnd(weights)
+
+        >>> np.allclose(first_call, second_call)
+        >>> False
+
+        This can be rectified by making the quantum node **immutable**.
+
+        .. code-block:: python
+
+            @qml.qnode(dev, mutable=False)
+            def circuit_rnd(weights):
+                RandomLayers(weights=weights, wires=range(2), seed=np.random.randint(100))
+                return qml.expval(qml.PauliZ(0))
+
+            first_call = circuit_rnd(weights)
+            second_call = circuit_rnd(weights)
+
+        >>> np.allclose(first_call, second_call)
+        >>> True
     """
     if seed is not None:
         np.random.seed(seed)
```

Comment:
(Outside the scope of this PR) I wonder if there is a way to locally set the numpy seed?

e.g.
- saving the global state with [get_state](https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.random.get_state.html)
- changing the seed
- at the end of the function, resetting the state to the saved value in the first step

Since the seed has a default value of 42, the global seed can be set by default, and potentially not noticed by the user.

Reply:
I'm not sure what you mean by "state"? Can you describe what changes in the behaviour this could entail?

I totally agree, randomness in qnodes is not a resolved issue...

Reply:
Great! Yes, I've not looked into it too much but I think the random number generator has a state. Setting the seed results in a new state being generated, but it's hard to recover the value of a seed from the state (see https://stackoverflow.com/questions/32172054/how-can-i-retrieve-the-current-seed-of-numpys-random-number-generator).

So I'd imagine something like
```python
def RandomLayers(...)
    state = np.random.get_state()

    if seed is not None:
        np.random.seed(seed)

    # Do circuit

    np.random.set_state(state)  # at end of function
```

Reply:
But not a worry for this PR :smile: